File: Week1/credit.c
#include <stdio.h>

int countDigits(unsigned long num);

int main(void) {
  unsigned long cardNum = 0;
   do {
       printf("Number: ");
       scanf("%lu", &cardNum);
   } while (countDigits(cardNum) < 13 || countDigits(cardNum) > 16);
    
  int sumDig = 0, otherDig = 0, firstTwo = 0;
  int len = countDigits(cardNum);
  int i = 0;
    
  while (cardNum != 0) {
    int lastDig = cardNum % 10;
    if (i % 2 == 0) { 
      otherDig += lastDig;
    }
    else {
      if (lastDig >= 5) {
        sumDig += ((lastDig - 4) * 2 - 1);
      }
      else {
        sumDig += lastDig * 2;
      }
    }
    if (countDigits(cardNum) == 2) {
      firstTwo += cardNum;
    }
    cardNum /= 10;
    i++;
  }
    
  if ((sumDig + otherDig) % 10 != 0) {
      printf("INVALID\n");
      return 0;
  }
  if ((firstTwo == 51 || firstTwo == 52 || firstTwo == 53 || firstTwo == 54 || firstTwo == 55) && (len == 16)) {
    printf("MASTERCARD\n");
  }
  else if ((firstTwo == 34 || firstTwo == 37) && (len == 15)) {
    printf("AMEX\n");
  }
  else if ((firstTwo / 10 == 4) && (len == 13 || len == 16)) {
    printf("VISA\n");
  }
  else {
    printf("INVALID\n");
  }
  return 0;
}

 
int countDigits(unsigned long num) {
  int count = 0;
  while (num != 0) {
    num /= 10;
    count++;
  }
  return count;
}

File: Week6/credit.py
def is_valid_card(card_num) -> boolean:
  if int(len(card_num)) not in [13, 15, 16]:
    return False
  temp = 0
  for i, digit in enumerate(card_num[::-1]):
    if not i % 2:
      temp += int(digit)
    else:
      temp += sum(int(x) for x in str(int(digit)*2))
  if str(temp)[-1] != '0':
    return False
  return True


def which_card(card_num):
  length = len(card_num)
  if length == 15 and int(card_num[:2]) in [34, 37]:
    print("AMEX")
  elif length == 16 and int(card_num[:2]) in range(51, 56):
    print("MASTERCARD")
  elif length in [13, 16] and int(card_num[0]) == 4:
    print("VISA")
  else:
    print("INVALID")


def main():
  card_num = input("Number: ")
  if not is_valid_card(card_num):
    print("INVALID")
  else:
    which_card(card_num)
    
if __name__ == "__main__":
  main()
